### PR TITLE
[Perf][Wan2.2] Skip attention mask for zero-padded SP sequences to avoid varlen path

### PR DESCRIPTION
**Pull request by jjuvonen-amd**
## Purpose

Reduce attention overhead for Wan2.2 T2V / I2V / TI2V / VACE workloads running Ulysses SP at resolutions where the video token count is not divisible by `ulysses_degree`. In that case `auto_pad` pads hidden states with zeros and the transformer constructed a boolean mask, routing every attention layer through the varlen path (unpad->kernel->repad). Introduce `DiffusionParallelConfig.mask_sp_padding` (default `False`) to skip this mask avoiding unpad/repad overhead of the varlen path. 

Since `_shard_with_auto_pad` always pads with zeros, the dense attention path produces nearly equivalent output. The original masked behavior is preserved when `mask_sp_padding=True`.

Some numerical difference is expected when skipping the attention mask, but this is negligible for the output video quality. Without the mask, padding tokens participate in softmax normalization, and the varlen and dense attention kernels may produce slightly different floating-point results for the same inputs.

`DiffusionParallelConfig` is used across all diffusion models, and the same attention mask pattern can be applied to other models (GLM Image, HunyuanVideo 1.5, Flux2 Klein, Qwen Image). However, to limit review-surface, this PR is focused on Wan2.2 optimization.

## Configuration

To restore original strict masking for numerically exact results, set `parallel_config.mask_sp_padding=true`. The new default is `False`, and this logs a warning message once if unmasked padding tokens occur during inference.

**YAML config**

```yaml
parallel_config:
  ulysses_degree: 8
  mask_sp_padding: false  # default is false - set true to restore strict masking
```

**Python**

```python
from vllm_omni.diffusion.data import DiffusionParallelConfig

parallel_config = DiffusionParallelConfig(
    ulysses_degree=8,
    mask_sp_padding=False,  # default is False - set True to restore strict masking
)
```

## Test Plan

Two new tests added to `test_diffusion_config_propagation.py` to verify the new parallel_config `mask_sp_padding` parameter survives serialization and defaults to `False`.

Run Wan2.2-14B-I2V end-to-end generation (8× GPU, `ulysses_degree=8`) with resolution configuration that doesn't split evenly across GPUs. Test without and with `mask_sp_padding=True` (original behavior) and visually compare outputs. Verify from profiles or logs that varlen path isn't used by default.

## Test Result

With `mask_sp_padding=True`, attention kernel log shows varlen path invoked on every attention
layer (`fmha_v3_varlen_fwd` on ROCm / aiter).

After: attention kernel log shows the dense attention path throughout.
The unpad/repad overhead of the varlen path is eliminated.

Output videos were compared using PSNR (average: 26.4 dB, min: 22.4 dB).
The numerical difference is expected: without the mask, padding tokens
participate in softmax normalization, and the varlen and dense attention
kernels may produce slightly different floating-point results for the same
inputs. Visual inspection confirms the outputs are of equivalent quality.
Note that bf16 diffusion sampling is sensitive to floating-point operation
order, so some divergence between runs is expected even without this change.

Original Wan2.2-14B-I2V output video:
https://github.com/user-attachments/assets/c549b75d-ccae-4eb1-9765-438129c17d02

Output video after changes:
https://github.com/user-attachments/assets/ae33342f-9f80-4497-9668-761e8b23bdd6

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by RuixiangMa** [comment]
@jjuvonen-amd Can you add before/after e2e latency or attention kernel execution time?

**Comment by jjuvonen-amd** [comment]
> @jjuvonen-amd Can you add before/after e2e latency or attention kernel execution time?

Thank you for looking into this @RuixiangMa. I observed **6% speed-up in Wan2.2-14B-I2V 8GPU e2e latency** with ROCm/aiter flash_attention, but the uplift from switching varlen to dense attention may vary across attention backends.

**Review by hsliuustc0106 — COMMENTED** [review]
LGTM - well-documented optimization with PSNR comparison and test coverage.

**Comment by hakankiymaz-amd** [comment]
I have tested Wan2.2-14B-I2V 8GPU Ulysses-degree=8, with torch.SDPA attention backend=cuDNN .
_**Single kernel execution time:**_
_before_ applying PR: `22ms`
_after_ applying PR: `10ms`
Kernel execution time performance speed up: `54%`

**Comment by hsliuustc0106** [comment]
@bjf-frz PTAL, cc @fhfuih @spencerr221

**Comment by hsliuustc0106** [comment]
thanks for your contribution, so this optmization could be extended to many other model as well, right?

**Comment by jjuvonen-amd** [comment]
> so this optimization could be extended to many other model as well, right?

@hsliuustc0106, yes, there are similar zero-pad attention mask patterns at least in GLM Image, HunyuanVideo 1.5, Flux2, and  Qwen Image. This PR covers only Wan models, but the same optimization could be extended to other models.

**Comment by hsliuustc0106** [comment]
> > so this optimization could be extended to many other model as well, right?
> 
> @hsliuustc0106, yes, there are similar zero-pad attention mask patterns at least in GLM Image, HunyuanVideo 1.5, Flux2, and Qwen Image. This PR covers only Wan models, but the same optimization could be extended to other models.

will you open new PRs for the rest?

**Comment by jjuvonen-amd** [comment]
> > > so this optimization could be extended to many other model as well, right?
> > 
> > 
> > @hsliuustc0106, yes, there are similar zero-pad attention mask patterns at least in GLM Image, HunyuanVideo 1.5, Flux2, and Qwen Image. This PR covers only Wan models, but the same optimization could be extended to other models.
> 
> will you open new PRs for the rest?

Yes, I'll open a follow-up PR with same optimizations to rest of the models once this PR gets merged. I left other models out from here to limit the testing surface.

Edit: There's already one follow-up PR open for [Qwen](https://github.com/vllm-project/vllm-omni/pull/4293) by @kTorp

**Comment by hsliuustc0106** [comment]
@tjtanaa can you help check why amd-ci fails again

**Review by hsliuustc0106 — APPROVED** [review]
lgtm

**Comment by tjtanaa** [comment]
Thanks @hsliuustc0106  I plan to look at it tomorrow. 🙏 qwen3 test started failing after one of the commits a few days ago.

**Comment by bjf-frz** [comment]
This may introduce a small numerical difference in attention outputs because padding tokens are no longer masked from the softmax, but it avoids the unpad/repad overhead caused by FlashAttention taking the varlen path. This looks like a worthwhile tradeoff, and the option to keep masking the padding is preserved.

  I would suggest adding a warning to make users aware of the possible numerical/quality difference when `mask_sp_padding=False`. It would also be helpful to document in the PR description how to configure `mask_sp_padding`, for example via `parallel_config.mask_sp_padding=true` to preserve the original masked behavior.

**Comment by hsliuustc0106** [comment]
> This may introduce a small numerical difference in attention outputs because padding tokens are no longer masked from the softmax, but it avoids the unpad/repad overhead caused by FlashAttention taking the varlen path. This looks like a worthwhile tradeoff, and the option to keep masking the padding is preserved.
> 
> I would suggest adding a warning to make users aware of the possible numerical/quality difference when `mask_sp_padding=False`. It would also be helpful to document in the PR description how to configure `mask_sp_padding`, for example via `parallel_config.mask_sp_padding=true` to preserve the original masked behavior.

make sense, acc is the highest level for production usage

**Comment by jjuvonen-amd** [comment]
> This may introduce a small numerical difference in attention outputs because padding tokens are no longer masked from the softmax, but it avoids the unpad/repad overhead caused by FlashAttention taking the varlen path. This looks like a worthwhile tradeoff, and the option to keep masking the padding is preserved.
> 
> I would suggest adding a warning to make users aware of the possible numerical/quality difference when `mask_sp_padding=False`. It would also be helpful to document in the PR description how to configure `mask_sp_padding`, for example via `parallel_config.mask_sp_padding=true` to preserve the original masked behavior.

Thank you for review and suggestions @bjf-frz
I added a one-time warning message (`logger.warning_once`) for when padding was applied but `mask_sp_padding=False`. The PR description now documents how to restore the original strict mask behavior.
